Custom type guard function

tinganho opened the pull request:
Implements only the function part of #1007. I thought it was best to have some early feedback first before I implement the class method part. So that I know I'm on the same page.

I consider the type guard type as a new type. Pls. check my tests for all specifications right now. I'm not sure if it's according to your specs. So would be good to have feedback on that.

Another question I have is if I should ban declaring type guard types on non-function-like declarations? I haven't implemented this ban yet. Example:
```typescript
var a: t is item //Error
```

mhegazy commented on src/compiler/binder.ts:
```diff
@@ -429,6 +431,9 @@ module ts {
                     bindDeclaration(<Declaration>node, SymbolFlags.TypeParameter, SymbolFlags.TypeParameterExcludes, /*isBlockScopeContainer*/ false);
                     break;
+                case SyntaxKind.TypeGuardType:
+                    bindDeclaration(<Declaration>node, (SymbolFlags.TypeGuardType | SymbolFlags.Value), SymbolFlags.TypeGuardTypeExcludes, /*isBlockScopeContainer*/ false);
```
why `SymbolFlags.Value`?

mhegazy replied:
why is that even bound? it does not create a symbol.. it is not creating a new type, it is just `boolean` with additional properties.. correct?

tinganho replied:
I removed the binding on a later commit.

mhegazy commented on src/compiler/binder.ts:
```diff
@@ -128,6 +128,8 @@ module ts {
                 case SyntaxKind.FunctionDeclaration:
                 case SyntaxKind.ClassDeclaration:
                     return node.flags & NodeFlags.Default ? "default" : undefined;
+                case SyntaxKind.TypeGuardType:
```
Do not like the name of the node/kind; i am not good with naming things though:) may be @JsonFreeman would have ideas.. possibly something with a "Specifier" or "narrowing" in the name would be better.

mhegazy commented on src/compiler/parser.ts:
```diff
@@ -1858,7 +1861,36 @@ module ts {
+                nextToken();
+            }
+            if (parseExpected(SyntaxKind.IsKeyword)) {
+                if (token === SyntaxKind.OpenBraceToken) {
```
i would parse a type here, then check it later in the checker to not be anonymous.

tinganho replied:
I've tried it before, I don't know why I get this error though:
```
    function isMissingTypeInTypeGuardType(x): x is {
             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
!!! error TS2391: Function implementation is missing or not immediately following the declaration.
        return true;
        ~~~~~~
!!! error TS1131: Property or signature expected.
    }
    ~
!!! error TS1128: Declaration or statement expected.
```
Instead of:
```
    function isMissingTypeInTypeGuardType(x): x is {
                                                   ~
!!! error TS1219: Type declaration expected.
        return true;
    }
```

Though I can admit checking for an open brace is not so good. Any good alternatives?

JsonFreeman replied:
Is this even correct behavior? If I wrote
```ts
function hasAnonymousReturn(x): x is { prop: string } {
}
```
This would be an error. But shouldn't it be allowed?

mhegazy commented on src/compiler/types.ts:
```diff
@@ -572,6 +574,11 @@ module ts {
+    
+    export interface TypeGuardTypeNode extends TypeNode {
+        target?: Identifier;
+        type: TypeNode;
```
I think this should be TypeReferenceNode,

tinganho replied:
Isn't TypeReferenceNode for Reference only? How about primitive types?

JsonFreeman replied:
Yes, I believe TypeNode is indeed what you want

mhegazy commented on src/compiler/types.ts:
```diff
@@ -572,6 +574,11 @@ module ts {
     }
+    
+    export interface TypeGuardTypeNode extends TypeNode {
+        target?: Identifier;
```
how about paramterName

mhegazy replied:
also it is kind of strange that missing means 'this'...

JsonFreeman replied:
Yeah I'd rather have target be of type `Identifier | ThisKeyword`. Oh, but ThisKeyword is not a node... I don't have a better idea

mhegazy commented on src/compiler/types.ts:
```diff
@@ -169,6 +170,7 @@ module ts {
         ConstructSignature,
         IndexSignature,
         // Type
+        TypeGuardType,
```
you would want to handle this in the declaration emitter as well.

tinganho replied:
I don't understand this. There is nothing new to emit?

JsonFreeman replied:
An alternative name that I think I like better is TypePredicate or PredicateType

mhegazy commented on src/compiler/checker.ts:
```diff
@@ -3690,6 +3697,48 @@ module ts {
+        function getTypeGuardType(node: TypeGuardTypeNode): Type {
+            let type = <TypeGuardType>createType(TypeFlags.TypeGuard);
+            let hasParameters = false;
+            type.parameterIndex = -1;
```
why not just store the paramterSymbol instead. it should be fairly simple when narrowing to check if this is the symbol being narrowed by checking identity.

tinganho replied:
I'm not sure I understand? I'm using the parameterIndex right now to get the argument name from a call expression. How would a parameterSymbol help me?

mhegazy commented on src/compiler/checker.ts:
```diff
@@ -3690,6 +3697,48 @@ module ts {
+                }
+            }
+            else {
+                type.type = anyType;
```
unknownType?

tinganho replied:
Fixes on a later commit

mhegazy commented on src/compiler/checker.ts:
```diff
@@ -3690,6 +3697,48 @@ module ts {
+            else {
+                type.type = anyType;
+            }
+            type.symbol = node.symbol;
```
i do not think this is correct.. we should not have a symbol here.

tinganho replied:
I removed it.

mhegazy commented on src/compiler/checker.ts:
```diff
@@ -3690,6 +3697,48 @@ module ts {
+            let hasParameters = false;
+            type.parameterIndex = -1;
+            if (node.target) {
+                let functionDeclaration = <SignatureDeclaration>node.target.parent.parent;
```
nothing says that this is a function declaration. you should be checking this earlier, and giving an error if a typeGuardType is used anywhere else.

tinganho replied:
I'm not sure I understand here either? This block of code is for checking that the declared TypeGuardType has a target that corresponds to a declared parameter in a signature.

I could change the variable name to just `signature`.

mhegazy commented on src/compiler/checker.ts:
```diff
@@ -4048,8 +4103,28 @@ module ts {
                 }
                 let saveErrorInfo = errorInfo;
+                if (source.flags & TypeFlags.TypeGuard && target.flags & TypeFlags.TypeGuard) {
+                    if (result = isRelatedTo((<TypeGuardType>source).type, (<TypeGuardType>target).type, reportErrors, headMessage)) {
```
i do not think this is correct...

```ts
declare function isFirstParamA (a, b): a is A;
declare function isSecondParamA (a, b): b is A;
declare function returnBoolean(a, b):boolean;

returnBoolean = isFirstParamA; // OK
isFirstParamA = returnBoolean; // Error
isFirstParamA = isSecondParamA ; // Error
```

tinganho replied:
I could add a check for the parameter index, to ensure they are the same?

tinganho replied:
Fixed now:
```
    var test1: (a, b) => boolean;
    test1 = function(a, b): a is A  {
        return true;
    };
    
    var test2: (a, b) => a is A;
    test2 = function(a, b): boolean  {
    ~~~~~
!!! error TS2322: Type '(a: any, b: any) => boolean' is not assignable to type '(a: any, b: any) => a is A'.
!!! error TS2322:   Type 'boolean' is not assignable to type 'a is A'.
        return true;
    };
    
    var test3: (a, b) => a is A;
    test3 = function(a, b): b is A  {
    ~~~~~
!!! error TS2322: Type '(a: any, b: any) => b is A' is not assignable to type '(a: any, b: any) => a is A'.
        return true;
    };
```

mhegazy commented on src/compiler/checker.ts:
```diff
@@ -3690,6 +3697,48 @@ module ts {
+                }
+                type.parameterName = node.target.text;
+            }
+            if (hasParameters && type.parameterIndex === -1) {
```
what about the `this` case.. you need to check that it is used in a correct way, i.e. only within a class maybe..

tinganho replied:
I will handle `this` later. I describes in my PR that I only handle the function case. Though it should be fairly simple to add the other case as well.

JsonFreeman replied:
This can be inside the `if (signature.parameters)` clause above, no?

mhegazy commented on src/compiler/checker.ts:
```diff
@@ -3690,6 +3697,48 @@ module ts {
+            }
+            if (node.type) {
+                type.type = getTypeFromTypeNode(node.type);
+                if (type.type.flags & TypeFlags.TypeGuard) {
```
i think this should be handled above, by only allowing typeGuardTypes in return type position of a function/method

JsonFreeman replied:
This only handles a type guard directly inside a type guard. If it is nested more deeply, I guess it will not be caught.

mhegazy commented on src/compiler/checker.ts:
```diff
@@ -3880,6 +3931,10 @@ module ts {
 
         function instantiateType(type: Type, mapper: TypeMapper): Type {
             if (mapper !== identityMapper) {
+                if (type.flags & TypeFlags.TypeGuard) {
```
i would move that to the bottom of the list

tinganho replied:
Fixed in a later commit.

mhegazy commented on src/compiler/checker.ts:
```diff
@@ -5437,11 +5517,32 @@ module ts {
+                        if (getSymbolAtLocation(expr.arguments[(<TypeGuardType>callType).parameterIndex]) === symbol) {
+                            let narrowedType = (<TypeGuardType>callType).type;
+                            if (isTypeAssignableTo(narrowedType, type)) {
+                               return narrowedType;
```
We should only narrow if the type is a subtype of the original type..

you also want to filter in case type is a union type.. 

```ts
 if (targetType) {
                    // Narrow to the target type if it's a subtype of the current type
                    if (isTypeSubtypeOf(narrowedType , type)) {
                        return targetType;
                    }
                    // If the current type is a union type, remove all constituents that aren't subtypes of the target.
                    if (type.flags & TypeFlags.Union) {
                        return getUnionType(filter((<UnionType>type).types, t => isTypeSubtypeOf(t, narrowedType )));
                    }
                }
```

mhegazy replied:
you would also want to handle the case of this. or make it an error in the parser for now. i think we can enable this in a different change.

mhegazy commented on tests/baselines/reference/typeGuardFunction.js:
```diff
@@ -0,0 +1,298 @@
+    return true;
+}
+
+function isArrayOf_2<T>(of: (item) => item is T, a: any[]): a is T[] {
```
you also want to use the new type construct as an inference target, e.g.: 
```ts
declare function getObjectOfType<T>(of: (item: any) => item is T, obj: any]): T;

// We can then infer T from the is clause
var f: string = getObjectOfType(isFoo, obj); // Error:  Type Foo is not assignable to string
```

tinganho replied:
I will add a test on that on a later commit.

mhegazy commented:
pinging @JsonFreeman and @ahejlsberg for this one as well.

mhegazy commented:
@tinganho thanks for submitting this change. I took a first stab at reviewing it. Overall i do not have major concerns, just my comments above. i would want @JsonFreeman and/or @ahejlsberg to take a look as well though.

JsonFreeman commented:
I haven't started reviewing this yet, but I don't think it's possible to only allow a certain type kind in return position because of type argument inference. You could have a function like this
```ts
function moveReturnToParameter(func: () => T): (p: T) => void;
```
and then passing in a type guarding function would output a function that takes a type guard as a parameter.

tinganho commented:
>```typescript
function moveReturnToParameter<T>(func: () => T): (p: T) => void;
```

Also a function that takes a type guard type(**not a type guard function**) as a parameter is pretty much useless.

Let say that that the returned function is:
```typescript
function(isFoo: item is Foo) {
  if(isFoo) {// This won't be able to take an argument so it is not useful?
  }
}
```

JsonFreeman commented:
Actually, I take that back. It is possible if you special case inference to infer boolean every time it encounters one of these type guards.

tinganho commented:
@JsonFreeman just for clarification do you mean that:
Passing in a type guard function to the following function
```typescript
function moveReturnToParameter<T>(func: () => T): (p: T) => void;
```
will convert the returned function to accept a boolean instead of a type guard type?

And so every non-function declaration with a declared type guard type will be converted to boolean?

tinganho commented:
@mhegazy I fixed some of the issue you pointed out. Though I didn't understood some of your comments, it would be good if you could clarify some of them.

JsonFreeman commented on src/compiler/types.ts:
```diff
@@ -1486,6 +1494,12 @@ module ts {
+    
+    export interface TypeGuardType extends Type {
+        parameterName: string;
+        parameterIndex?: number; // Call expression argument
```
If the parameterName is a name found in a binding pattern, will parameterIndex be the index of the top level binding pattern? How will you then know which binding element needs to be narrowed? Also, please ensure you have tests that try to type guard a parameter inside a binding pattern.

JsonFreeman commented on src/compiler/diagnosticMessages.json:
```diff
@@ -683,6 +683,14 @@
+        "category": "Error",
+        "code": 1219
+    },
+    "Can't define a type-guard type inside a type-guard type.": {
```
Will we eventually only allow a type guard type (type predicate) in return position? If so, I think this restriction is just a special case of that, no?

I would use "cannot" instead of "can't". Just for style.

JsonFreeman commented on src/compiler/parser.ts:
```diff
@@ -1858,7 +1861,36 @@ module ts {
+        function parseTypeGuardType(parameterList?: NodeArray<ParameterDeclaration>): TypeGuardTypeNode {
+            let node = <TypeGuardTypeNode>createNode(SyntaxKind.TypeGuardType);
+            if (token === SyntaxKind.Identifier) {
+                node.target = createIdentifier(true);
```
Use a named parameter comment to label the `true` literal.

JsonFreeman commented on src/compiler/parser.ts:
```diff
@@ -1858,7 +1861,36 @@ module ts {
+        }
+        
+        function isStartOfTypeGuardType(): boolean {
+            return (token === SyntaxKind.Identifier || token === SyntaxKind.ThisKeyword) && lookAhead(nextTokenIsIsKeyword);
```
If you see a ThisKeyword, I think you can be certain you have a type guard.

JsonFreeman commented on src/compiler/parser.ts:
```diff
@@ -2313,6 +2345,9 @@ module ts {
         }
 
         function parseTypeWorker(): TypeNode {
+            if (isStartOfTypeGuardType()) {
```
This will do a look ahead for almost every type node. I would not do it this way. Instead, I'd have a function called parseTypeGuardTypeOrHigher, that does something like this:
```ts
let type = parseUnionTypeOrHigher();
while (token === SyntaxKind.IsKeyword) {
    // create a new TypeGuardTypeNode 'node'
    node.target = type;
    // more stuff
    type = node;
}
return type;
```
Basically what parseUnionTypeOrHigher does, but for type guard types instead.

I am assuming here that you want type guard types to be the loosest precedence, so that
```ts
x is Foo | Bar
```
means that the parameter x is guarded to be of type `Foo | Bar`.

JsonFreeman commented on src/compiler/checker.ts:
```diff
@@ -4057,12 +4111,27 @@ module ts {
                 let saveErrorInfo = errorInfo;
+                if (source.flags & TypeFlags.TypeGuard && target.flags & TypeFlags.TypeGuard) {
+                    if ((<TypeGuardType>source).parameterIndex &&
+                        (<TypeGuardType>target).parameterIndex &&
```
If either parameterIndex is 0, this will pass. I think you mean `parameterIndex >= 0`

JsonFreeman commented on src/compiler/checker.ts:
```diff
@@ -5450,11 +5524,32 @@ module ts {
+            
+            function narrowTypeByCallExpression(type: Type, expr: CallExpression, assumeTrue: boolean): Type {
+                // Check that type is not any, assumed result is true.
+                if (type.flags & TypeFlags.Any || !assumeTrue) {
```
If assumeTrue is false and type is a union, I think you can remove constituents that are subtypes of the guard. That is what we do for instanceOf guards.

JsonFreeman commented on src/compiler/checker.ts:
```diff
@@ -3690,6 +3697,47 @@ module ts {
+                if (signature.parameters) {
+                    for (let i = 0; i < signature.parameters.length; i++) {
+                        let param = signature.parameters[i];
+                        if (param.name.kind === SyntaxKind.Identifier && (<Identifier>param.name).text === node.target.text) {
```
Ah, so it doesn't find a name inside a binding pattern. I think that is fine.

JsonFreeman commented on src/compiler/checker.ts:
```diff
@@ -9028,8 +9123,19 @@ module ts {
+                    checkIfNonVoidFunctionHasReturnExpressionsOrSingleThrowStatment(node, getTypeFromTypeNode(node.type));
+                }
+                let type = getTypeFromTypeNode(<TypeGuardTypeNode>node.type);
+                if (type.flags & TypeFlags.TypeGuard && ~(<TypeGuardType>type).parameterIndex) {
```
Please check `>= 0` instead of `~`

JsonFreeman commented on src/compiler/checker.ts:
```diff
@@ -9028,8 +9123,19 @@ module ts {
+                        (<TypeGuardType>type).type,
+                        getTypeAtLocation(node.parameters[(<TypeGuardType>type).parameterIndex]),
+                        (<TypeGuardTypeNode>node.type).type, 
+                        Diagnostics.Type_0_is_not_assignable_to_type_1
```
You can leave off the message here. I believe the one you passed is the default message.

JsonFreeman replied:
Also, I think all this checking should really be done in checkSignatureDeclaration. Function expressions for instance do not get here.

JsonFreeman commented on src/compiler/checker.ts:
```diff
@@ -4929,6 +4998,12 @@ module ts {
+                }
+                if (source.flags & TypeFlags.TypeGuard) {
+                    source = (<TypeGuardType>source).type;
+                }
```
This does not seem correct.
```ts
declare function isFoo(x): x is Foo;
declare function generic<T>(func: (p) => T): T;
generic(isFoo);
```
Here you will infer Foo for T, but really I think you should infer boolean.

JsonFreeman replied:
Also, does the following work?
```ts
declare function convert<T, U>(source: T, predicate: (x: T) => x is U): U;
declare function isNumber(x: any): x is number;
let x: any;
convert(x, isNumber); // should return number
```

tinganho replied:
I think I miss this generic cases.

tinganho replied:
>```typescript
declare function isFoo(x): x is Foo;
declare function generic<T>(func: (p) => T): T;
generic(isFoo);
```
Here you will infer Foo for T, but really I think you should infer boolean.

Does this infer a boolean because `x is Foo` is not assignable to `T`?

So we can't do things like this?:

```typescript
declare function isFoo(x): x is Foo;
declare function generic<T>(a: Foo, func: (p) => T): T;
if(generic(a, isFoo)) {
  a;// type Foo
}
```

JsonFreeman replied:
It infers a boolean because `x is Foo` is of type boolean. It is not a question of assignability to T, you are actually inferring what type T should be. And in this case it is the type given by `x is Foo` which is boolean.

For your examples, you are correct, they should not work as type guards. The reason is that in both of them `x is Foo` is actually a statement about `p` in the context of the call to `generic`. And there is no way to know what `generic` passes for `p` internally (as an argument to `func`). In particular, we have no idea if the argument bears any relation to `a`, which is the thing you're trying to narrow.

JsonFreeman commented:
Hmm, thinking about this more, I don't really know what it means for a type guard type to flow through the type system. Once a type guard type gets separated from a call to its originating signature, it becomes kind of meaningless. If these are truly to be their own kind of types, I would expect the following to work:
```ts
declare function isString(x): x is string;
var v: string | number;
var vIsString = isString(v);
if (vIsString) {
    // Here v has type string
}
```
If this works, then that's pretty cool, and maybe we should keep it. I was more thinking that these type nodes would just become booleans, but if you are in the process of narrowing, you can get a type predicate from the signature that the call expression is calling.

I think we should try to determine if making type guards their own types will cause anything weird to happen in the type system. If it doesn't, then I am fine with this. Some of the inference cases I have been bringing up are examples of some things to watch out for.

JsonFreeman commented:
Actually, my example above will not work because `vIsString` is not a call expression. Given that, is there a benefit to having a separate kind of type, that I am missing? I feel like the boolean type approach would be simpler.

tinganho commented:
@JsonFreeman How will you store the type predicate? Storing it on the `booleanType`? or just the `signature`? Is there a free ride we can have by just using booleans instead?

JsonFreeman commented:
You would store it on the signature.

JsonFreeman commented:
Talked with @mhegazy, and we thought of something. We think it does not make sense to treat a type predicate as a type in the type system, because it is an interpretation of the significance of a boolean, and its meaning is highly localized. It should be a property of the signature instead.

You can still check assignability of type predicates when you check assignability of signatures. When you compare signatures for assignability, you would check parameters, return type, and type predicate (if any). The signatures would be assignable if the type predicate has the same argument index, and the types of the type predicates are assignable.

You can also do type argument inference on the type predicate, but again, it is when you visit the signature. If both the source and the target have type predicates with the same parameter index, you can infer between the type portion of the type predicate.

And of course, instantiating a signature should instantiate the type of its type predicate, similar to what you have already added.

Let me know if you have concerns about any of these ideas, or questions about how/where to implement them.

JsonFreeman commented:
Again, the purpose of doing it this way is to prevent the type predicate from flowing anywhere.

tinganho commented:
@JsonFreeman thanks for your feedback very helpful. I will rework this PR then.

JsonFreeman commented:
I also think it should be an error (at least for now) to have a type predicate refer to a binding element name in a parameter list. Only top level parameter names (identifiers) should be fair game.

tinganho commented:
@JsonFreeman I will add an error for that. I guess I will show errors on these too:
```typescript
var a: a is Array;// Can only assign a type predicate to function declarations.
function(a: a is Array) {}; // Can only assign a type predicate to function declarations.
etc..
```

JsonFreeman commented:
Yes, I think type predicates should only be allowed in return type position. I think they should also be allowed on ambient declarations, and overloads.

jbondc commented:
My question is how would this be used in the compiler? 

Lots of cases where user-defined functions could cast/guard:
```
 // 1252: cast to ModuleDeclaration
 return (declaration.kind === SyntaxKind.ModuleDeclaration && (<ModuleDeclaration>declaration).name.kind === SyntaxKind.StringLiteral) ||

 // 376: cast to ClassDeclaration
 if (location.parent.kind === SyntaxKind.ClassDeclaration && !(location.flags & NodeFlags.Static)) {
    let ctor = findConstructorDeclaration(<ClassDeclaration>location.parent);

isKind(node: Node, kind: SyntaxKind): ? {
  return node.kind === kind;
}

// 1475: cast to IntrinsicType
if (type.flags & TypeFlags.Intrinsic) {
    // Special handling for unknown / resolving types, they should show up as any and not unknown or __resolving
     writer.writeKeyword(!(globalFlags & TypeFormatFlags.WriteOwnNameForAnyLike) &&
      (type.flags & TypeFlags.Any) ? "any" : (<IntrinsicType>type).intrinsicName);
}

isType(type: Type, flags: TypeFlags): ? {
  return node.flags & flags;
}
``` 
Seems to be two ways of thinking about it:
a) macro-like function guards (as above) & similar to patch
b) Fixing instanceof 

```
interface IntrinsicType {
} where  <instanceof(T) { return T.flags & TypeFlags.Intrinsic };

if (type instanceof IntrinsicType) { // emits type.flags & TypeFlags.Intrinsic
    // Special handling for unknown / resolving types, they should show up as any and not unknown or __resolving
     writer.writeKeyword(!(globalFlags & TypeFormatFlags.WriteOwnNameForAnyLike) &&
      (type.flags & TypeFlags.Any) ? "any" : type.intrinsicName);
}
```

DanielRosenwasser commented:
> My question is how would this be used in the compiler?

We have functions like

* `isDeclaration`
* `isBindingPattern`
* `isTemplateLiteralKind`

And others.

RyanCavanaugh commented:
Also, if we add "overload on enum" as a parallel to overload on string, we could have definitions like:
```ts
function isOfKind(n: Node, s: SyntaxKind.ThrowKeyword): n is ThrowKeyword;
function isOfKind(n: Node, s: SyntaxKind.ReturnKeyword): n is ReturnKeyword;
/* repeat ad infinitum... */
function isOfKind(n: Node, s: SyntaxKind): boolean;
```

JsonFreeman commented:
I like @RyanCavanaugh's idea

jbondc commented:
Like it too, though if you start doing:
```
function isOfKind(n: Node, s: SyntaxKind.ModuleDeclaration): n is ModuleDeclaration;
function isOfKind(n: Node, s: SyntaxKind.ClassDeclaration): n is ClassDeclaration;
```
You might as well attach the 'is' notion directly to the type declaration:
```
    export interface ModuleDeclaration extends Declaration, ModuleElement {
        name: Identifier | LiteralExpression;
        body: ModuleBlock | ModuleDeclaration;
    } is(n: Node) { return n.flags & SyntaxKind.ModuleDeclaration }

Or by reference:
type foo = ModuleDeclaration is (n: Node) { return n.flags & SyntaxKind.ModuleDeclaration }
```
Something like that. People are going to start asking for:
```
if(node is ModuleDeclaration) {
  //...
}
```

tinganho commented:
@JsonFreeman I just found out that if the type guard "target" is a property access expression — the type guard doesn't work.

```typescript
class A {
	propA;
}
class B {
	propB;
}
let a: {
	b: A | B;
}
let b = a.b instanceof B && a.b.propB; // Error propB doesn't exists on type 'A | B'
```
This is because it "walks" backwards to get the type of each element in a property access expression until it finds the first element(the identifier). Then it narrows the type of that one.

I don't really know how to fix this without introducing a variable before the "walking" occurs.

jbondc commented:
Actually:
```
function isOfKind(n: Node, flags: SyntaxKind) {
  return n.flags & flags;
}

// By reference:
type ModuleDeclarationKind = ModuleDeclaration is<T> => { return T.flags & SyntaxKind.ModuleDeclaration };
type ModuleDeclarationKindIdentityWithFunction = ModuleDeclaration is<T> => isOfKind(T, s: SyntaxKind.ModuleDeclaration);
```
I think you get the same result but 'ModuleDeclarationKindIdentityWithFunction' requires the emit of a function but 'ModuleDeclarationKind' would only emit inline the 'is function' body.

jbondc commented:
@tinganho Type guards only work on variables, see #1260

It's a hard one to solve, likely related to having some control flow #2854.
The challenge is finding a model that performs well.

jbondc commented:
From my understanding, a custom type guard function solves the performance problem :wink: 
Not sure if property access guards are needed.

jbondc commented:
A possible test case:
```
function getKind(node: Node): is VariableLikeDeclaration|SignatureDeclaration {
      switch(node.kind) {
            case SyntaxKind.Parameter:
            case SyntaxKind.PropertyDeclaration:
            case SyntaxKind.PropertySignature:
            case SyntaxKind.PropertyAssignment:
            case SyntaxKind.ShorthandPropertyAssignment:
            case SyntaxKind.VariableDeclaration:
            case SyntaxKind.BindingElement:
                 return 'VariableLike' is VariableLikeDeclaration; 

            case SyntaxKind.FunctionType:
            case SyntaxKind.ConstructorType:
            case SyntaxKind.CallSignature:
            case SyntaxKind.ConstructSignature:
            case SyntaxKind.IndexSignature:
                return 'Signature' is SignatureDeclaration;
     }
}

switch(getKind(node)) {
       case 'VariableLike': // node narrows to VariableLikeDeclaration
           return visitNodes(cbNodes, node.decorators) ||
                    visitNodes(cbNodes, node.modifiers) ||
                    visitNode(cbNode, node.propertyName) ||
                    visitNode(cbNode, node.dotDotDotToken) ||
                    visitNode(cbNode, node.name) ||
                    visitNode(cbNode, node.questionToken) ||
                    visitNode(cbNode, node.type) ||
                    visitNode(cbNode, node.initializer);
       case 'Signature': // node narrows to SignatureDeclaration
                return visitNodes(cbNodes, node.decorators) ||
                    visitNodes(cbNodes, node.modifiers) ||
                    visitNodes(cbNodes, node.typeParameters) ||
                    visitNodes(cbNodes, node.parameters) ||
                    visitNode(cbNode, node.type);
}
```
Would be neat if ~ that worked:
https://github.com/Microsoft/TypeScript/blob/master/src/compiler/parser.ts#L45

tinganho commented:
@jbondc I think #1260 is referring to property access type guards. What I meant was property access expressions together with instanceof/typeof/type-guard-functions.

JsonFreeman commented:
As @jbondc points out, properties are not narrowed, only identifiers. You are right that #1260 is for something different though, treating property accesses as a kind of type guard, versus enabling properties to be narrowed by a type guard.

To enable this though, we'd have to make sure that the type guard target is not only the same property as the one you are trying to narrow, but accessed off the same object. So the following would not be narrowed:
```ts
let a: {
    b: A | B
};
let a2: typeof a;
let b = a.b instanceof B && a2.b.propB; // Should not narrow
```
We would have to make sure all identifiers in the property access are the same as the one being narrowed.

This is easy to work around though, if you put `a.b` in a local first.

CyrusNajmabadi commented on src/compiler/parser.ts:
```diff
@@ -83,7 +83,8 @@ module ts {
                     visitNodes(cbNodes, (<SignatureDeclaration>node).parameters) ||
-                    visitNode(cbNode, (<SignatureDeclaration>node).type);
+                    visitNode(cbNode, (<SignatureDeclaration>node).type) ||
+                    visitNode(cbNode, (<SignatureDeclaration>node).typePredicate);
```
I would not do this.  "typePredicate" is just another way of specifying the type of the thing.  I would instead add ```interface TypePredicateNode : TypeNode { name: Identifier, type: TypeNode }```

tinganho replied:
I had a discussion earlier with @JsonFreeman . We came to the conclusion that type predicate info should be stored in a separate property in `Signature`. I thought to keep things consistent  we would store it on a separate property in `SignatureDeclaration`.

In my first implementation of this — I stored it on the type node.

JsonFreeman replied:
I was thinking it would be stored separately on the Signature, but not separately on the SignatureDeclaration. It is semantically separate from the return type, but syntactically not.

CyrusNajmabadi commented on src/compiler/parser.ts:
```diff
@@ -765,7 +770,7 @@ module ts {
         }
 
-        function parseErrorAtCurrentToken(message: DiagnosticMessage, arg0?: any): void {
+        function parseErrorAtCurrentToken(message: DiagnosticMessage, arg0?: any) {
```
Please don't remove this type annotation.

CyrusNajmabadi commented on src/compiler/parser.ts:
```diff
@@ -2011,6 +2016,34 @@ module ts {
             return parseInitializer(/*inParameter*/ true);
         }
 
+        function parseTypePredicate(signature: SignatureDeclaration) {
```
I would not do things this way.  Instead, just augment parseType to work as it does today.  Except that if you parse out a type that is just an identifier, and you see "is" as the current ```token```, then instead parse out a TypePredicateNode.  Then, in the checker if you ever see a TypePredicateNode *not* as the return type of an appropriate signature, then you error.

tinganho commented:
I think there are still some uncertainty whether we should treat type predicates as a new type or as an another signature property.

Here is just a copy paste of a comment I made in the parser that could be useful:

>@JsonFreeman suggested something similar. Though I think the problem here is that there are different views of the problem. Should we treat type predicates as types or not? Seems like you want the parser to treat it as an another type but the checker as an another signature property?

>I think to keep things consistent we should treat it both equal in the parser and checker. That means either as types or as a different signature property — a type predicate property.

>At first, I thought type predicate was an another type. Because it occupies the same syntax space as return types. But after some re-thoughts it makes sense to treat it as another signature property, since it already implies a boolean type. (Also if we treat it as an another signature property — I would suggest to change the colon token : to something else to minimize confusion.)

tinganho commented:
Seems like some comments where deleted after my rebase of master. Here are a link to them https://github.com/tinganho/TypeScript/commit/373a776a7021a4f11f79e5f8fe6ae3cead57de8b#commitcomment-11492134.

Sorry for any inconvenience I made.

tinganho commented:
I think if you click "Files changed" above and then makes some comments. The comments will not be deleted on rebase. Then you comment on the PR instead of commits in my fork. @CyrusNajmabadi comments are still on this thread.

CyrusNajmabadi commented:
@tinganho  I see what you're saying.  The more i think about it, the more i think that "x is string" is not the right syntax we want for this.  Effectively, in english, what this feature is saying is:

"here's a function which returns a boolean.  If that boolean is true, then it implies that some parameter 'p' has a specified type."  Or, simplified 

"A true return value means some parameter 'p' has a certain type"  Or, simplified

"true implies parameter 'p' has type"  or, simplfied

"true implies p is type"  or, simplified

```true -> p is type```

I'm not saying that this should be the final syntax.  But i'm definitely leaning toward something that is more explicit about what we're trying to say about this signature.  Things i like about syntax like this:

1. It is very obvious that the return type is actually a boolean.  "true" here acts similarly to our "string literal types" in signatures.  In this case though, "true" would be a boolean literal type.  
2. There is a strong conveyance of what it means when the result is "true", namely that it now means that some specified parameter is narrowed to the provided type.

Right now, i think that ```p is type``` is just a little too odd to see in the return type location.  Both ```p``` and the ```type``` look like types.  And yet neither of them is actually the real type of the function.

tinganho commented:
>```typescript
function isFoo(a: Foo | Bar): true -> a is Foo { }
```

:+1: This does make things clearer.

jbondc commented:
@CyrusNajmabadi Definitely agree with that thinking, for the decorator, wonder if it would be more useful on the return expression: 
```
function isFoo(a: Foo | Bar): true {
     @ReturnLiteralTypeNarrow(a,Foo)
     return true;
}
```
Think it means it work in switch statements?
```
function getKind(node: Node): string {
      switch(node.kind) {
            case SyntaxKind.Parameter:
            case SyntaxKind.PropertyDeclaration:
            case SyntaxKind.PropertySignature:
            case SyntaxKind.PropertyAssignment:
            case SyntaxKind.ShorthandPropertyAssignment:
            case SyntaxKind.VariableDeclaration:
            case SyntaxKind.BindingElement:
                 @ConstanValueNarrow(node,VariableLikeDeclaration)
                 return "VariableLike"; 

            case SyntaxKind.FunctionType:
            case SyntaxKind.ConstructorType:
            case SyntaxKind.CallSignature:
            case SyntaxKind.ConstructSignature:
            case SyntaxKind.IndexSignature:
                @ConstanValueNarrow(node,SignatureDeclaration)
                return "Signature";
     }
}
```
Or possibly:
```
function getKind(node: Node): "VariableLike" | "Signature" {
      switch(node.kind) {
            case SyntaxKind.Parameter:
            case SyntaxKind.PropertyDeclaration:
            case SyntaxKind.PropertySignature:
            case SyntaxKind.PropertyAssignment:
            case SyntaxKind.ShorthandPropertyAssignment:
            case SyntaxKind.VariableDeclaration:
            case SyntaxKind.BindingElement:
                 @ReturnLiteralTypeNarrow(node,VariableLikeDeclaration)
                 return "VariableLike"; 

            case SyntaxKind.FunctionType:
            case SyntaxKind.ConstructorType:
            case SyntaxKind.CallSignature:
            case SyntaxKind.ConstructSignature:
            case SyntaxKind.IndexSignature:
               @ReturnLiteralTypeNarrow(node,SignatureDeclaration)
               return "Signature";
     }
}
```
Not sure about implementation but what I like here is there's less notation fluff around the function.

jbondc commented:
Ahh this more elegant:
```
// Boolean enum
enum BooleanIdentityMap {
  FooType = true : Foo
}

function isFoo(a: Foo | Bar): BooleanIdentityMap {
     return BooleanIdentityMap.FooType;
}

// String enum
enum StringMapToType {
  [prop: string]: string;
  Signature: SignatureDeclaration,
  VariableLike: VariableLikeDeclaration,
}

function getKind(node: Node): StringIdentityMap {
      switch(node.kind) {
            case SyntaxKind.Parameter:
            case SyntaxKind.PropertyDeclaration:
            case SyntaxKind.PropertySignature:
            case SyntaxKind.PropertyAssignment:
            case SyntaxKind.ShorthandPropertyAssignment:
            case SyntaxKind.VariableDeclaration:
            case SyntaxKind.BindingElement:
                 return StringIdentityMap.Signature; 

            case SyntaxKind.FunctionType:
            case SyntaxKind.ConstructorType:
            case SyntaxKind.CallSignature:
            case SyntaxKind.ConstructSignature:
            case SyntaxKind.IndexSignature:
                 return StringIdentityMap.VariableLike; 
     }
}
```
An enum is a listing of constant values.
We can add a 'type hint' to an enum member, to map a constant value to a type.

JsonFreeman commented:
Please do not rebase once you have sent out a pull request.

JsonFreeman commented on src/compiler/checker.ts:
```diff
@@ -4632,6 +4659,43 @@ module ts {
+
+                if (source.typePredicate && target.typePredicate) {
+                    let hasDifferentParamaterIndex = source.typePredicate.parameterIndex !== target.typePredicate.parameterIndex;
+                    let hasDifferentTypes = !isTypeIdenticalTo(source.typePredicate.type, target.typePredicate.type);
```
If you don't report both errors (which I've already explained), then you can defer this until you know the indices are equal.

tinganho replied:
Fixed both problems. I hope you mean by defer by just adding another if statement to 
```typescript
if(hasDifferentParamaterIndex) {?
   let hasDifferentTypes... 
```

tinganho replied:
Oh I see what you mean now. Ended up with this instead:
```typescript
let hasDifferentTypes: boolean;
if (hasDifferentParamaterIndex || (hasDifferentTypes = !isTypeIdenticalTo(source.typePredicate.type, target.typePredicate.type))) {
```

JsonFreeman commented on src/compiler/checker.ts:
```diff
@@ -4632,6 +4659,43 @@ module ts {
+                        }
+                        return Ternary.False;
+                    }
+                    return Ternary.True;
```
Remove this

tinganho replied:
Fixed.

JsonFreeman commented on src/compiler/diagnosticMessages.json:
```diff
@@ -703,6 +703,31 @@
+        "category": "Error",
+        "code": 1224
+    },
+    "A type-guard function can only return a boolean.": {
```
Do you use this?

tinganho replied:
Sorry forgot to remove it.

tinganho replied:
Fixed.

JsonFreeman commented on src/compiler/diagnosticMessages.json:
```diff
@@ -703,6 +703,31 @@
         "category": "Error",
         "code": 1223
     },
+    "A non-type guard function is not assignable to a type guard function.": {
```
You can say something like `Signature '{0}' must have a type predicate`.

JsonFreeman commented:
I think this will be easier to review if you remove all the stuff having to do with `this` where the type predicate name is undefined.

tinganho commented:
@JsonFreeman I will remove all reference to them.

JsonFreeman commented:
Thanks @tinganho

JsonFreeman commented on src/compiler/checker.ts:
```diff
@@ -5164,6 +5228,16 @@ module ts {
+                        // we can't infer any type information from the return types. We can just add a return
+                        // statement after the below infer statement.
+                        inferFromTypes(source.typePredicate.type, target.typePredicate.type);
+                        return;
```
You can move this return into the outer `if` block

tinganho replied:
Fixed.

tinganho commented:
@JsonFreeman I reverted back to your original idea with `parseTypePredicateAndHigher` and addressed your feedback. I also removed all references to type predicate target `this`.

I removed the new node links. Turns out that we didn't need them.

Though I didn't have time to fix the overload of signature problem. Don't know yet how to address that.

JsonFreeman commented:
Thanks for all the iterating @tinganho. The signature overload problem is one that should be solved separately, not as part of this pull requests. We've known about it for a while, and need a more general solution anyway that is independent of type predicates.

JsonFreeman commented on src/compiler/checker.ts:
```diff
@@ -5561,30 +5627,69 @@ module ts {
                 return type;
             }
 
+            function getOptionalNarrowedType(originalType: Type, narrowedTypeCandidate: Type) {
```
Why optional?

tinganho replied:
Sometimes the type is not narrowed.

JsonFreeman replied:
I think everywhere the compiler uses the word "narrow", it implies that it might not happen.

JsonFreeman commented on src/compiler/checker.ts:
```diff
@@ -5561,30 +5627,69 @@ module ts {
+                if (!assumeTrue) {
+                    if (type.flags & TypeFlags.Union && 
+                        signature.typePredicate && 
+                        getSymbolAtLocation(expr.arguments[signature.typePredicate.parameterIndex]) === symbol) {
```
This looks very similar to shouldNarrowTypeByTypePredicate. Why is the false case handled in here, but the true case handled in shouldNarrowTypeByTypePredicate?

JsonFreeman replied:
I would make a decision for whether the symbol matches, without caring about assumeTrue. Then use assumeTrue to decide how to narrow.

tinganho replied:
The false case removes all constituent subtypes to the type predicate type if the narrow target type is of union type and assumed false.

I simplified it a bit according to your feedback.

JsonFreeman commented on src/compiler/checker.ts:
```diff
@@ -5561,30 +5627,69 @@ module ts {
+                        signature.typePredicate && 
+                        getSymbolAtLocation(expr.arguments[signature.typePredicate.parameterIndex]) === symbol) {
+
+                        return getUnionType(filter((<UnionType>type).types, t => !isTypeSubtypeOf(t, signature.typePredicate.type)));
```
Very similar to getOptionalNarrowedType.

tinganho replied:
Notice the negation `!`.

JsonFreeman replied:
Yes, I think that's the only difference

JsonFreeman commented on src/compiler/parser.ts:
```diff
@@ -2469,7 +2488,7 @@ module ts {
                 return parseFunctionOrConstructorType(SyntaxKind.ConstructorType);
             }
-            return parseUnionTypeOrHigher();
+            return parseTypePredicateOrHigher();
```
Yes, this is better.

JsonFreeman commented on src/compiler/diagnosticMessages.json:
```diff
@@ -703,6 +703,27 @@
+        "category": "Error",
+        "code": 1227
+    },
+    "Cannot define type predicate '{0}' as a type to a type predicate.": {
```
You need a more general error that will disallow type predicates in all positions, except return type. A nested type predicate is just a specific case of that.

tinganho replied:
How about "Type predicates are only allowed in return type position" ?

JsonFreeman replied:
That is good, but let's use the singular "A type predicate is"

JsonFreeman commented on src/compiler/parser.ts:
```diff
@@ -2461,6 +2464,22 @@ module ts {
+            let type = parseUnionTypeOrHigher();
+            if (token === SyntaxKind.IsKeyword &&
+                type.kind === SyntaxKind.TypeReference &&
+                (<TypeReferenceNode>type).typeName.kind === SyntaxKind.Identifier) {
```
You need to also check that the node has no type arguments. A `TypeReferenceNode` could have type arguments, and you don't want this to be parsed as a type predicate
```ts
Foo<T> is Bar
```

CyrusNajmabadi replied:
This approach is also not desirable because you end up having to extract out a child node of another node.  As mentioned before, you can change parseTypeReference and then make decision immediately after parsing teh entityName out.  It will make things simpler and push the check to the most immediate point possible.

tinganho replied:
I changed to @CyrusNajmabadi idea.

JsonFreeman replied:
Yeah I agree with the suggested approach.

JsonFreeman commented:
One more thing. You will need to check that the parameter referenced by the type predicate is not a rest parameter.

tinganho commented:
I just added a restriction on type predicates signature declarations. Constructors, get/set accessors can't have type predicates. I also fixed some of the feedback.

tinganho commented:
Just added an error for referencing a top-level binding element. Also addressed all the other feedback.

JsonFreeman commented:
Thanks @tinganho! After these last few minor tweaks, can you merge the upstream master branch into here? Then we can work on merging this in.

yuit commented:
Let me know if I miss anything here, but wonder what happens in the following example?

```ts
function isB(b: any): x is B {...}
```
I would assume we will give an error in this case.Also could you add the case in the tests.

jbondc commented:
This feature would be better simplified as 'type narrowing by constant value'.
```
function isB(b: any): boolean {
   const isBoo = true is B;
   return isBoo;
}
// Or with const enums 
const enum TrueThings {
    isB = true is B;
}
function isB2(b: any): boolean {
   return TrueThings.isBoo;
}
// Type narrowing by call expression that returns a type with an identity map of constants to types:
class C implements B {
}
let c = new C;
if(isB(c)) {
   // Function signature has an identityMap["true"] = C and C is assignable to identityMap["true"] ;
   c; // narrowed to B
} 
// Type narrowing by property access:
class D implements B {
   isBoo = TrueThings.isBoo;
}
let d = new D;
if(d.isBoo) {
   // the constant value has an identityMap["true"] = C and C is assignable to identityMap["true"], narrow to B ;
} 
```

JsonFreeman commented on src/compiler/checker.ts:
```diff
@@ -8486,6 +8587,34 @@ module ts {
         }
 
+        function getTypePredicateParameterIndex(parameterList: NodeArray<ParameterDeclaration>, parameter: Identifier): number {
+            let index = -1;
```
Is index used?

JsonFreeman commented on src/compiler/checker.ts:
```diff
@@ -8486,6 +8587,34 @@ module ts {
+                        return i;
+                    }
+                }
+            }
```
`return -1`, right?

JsonFreeman commented:
Yes @yuit, that code would be an error. The function `getTypePredicateParameterIndex` would return -1 / undefined, and then we would issue the error at https://github.com/Microsoft/TypeScript/pull/3262/files#diff-c3ed224e4daa84352f7f1abcd23e8ccaR8680

yuit commented:
@JsonFreeman. I see, Thanks :)

JsonFreeman commented:
Awesome job @tinganho! Thank you for implementing this much requested feature!

RyanCavanaugh commented:
Really cool to see this feature go in. Great work! :trophy:

tinganho commented:
thanks @JsonFreeman @RyanCavanaugh . :smiley:

DickvdBrink commented:
I like this! thanks @tinganho

ahejlsberg commented:
@tinganho Nice work!

RichiCoder1 commented:
*prepares to re-write all the definition files*

Nice work @tinganho!
